Extract daemon configuration and discovery to their own package

vdemeester opened the pull request:
The `daemon` package is really big and not well unit tested. This small step extract 2 relatively less coupled functionnality of `daemon` into their own package.

This also moves some cli specific in `cmd/dockerd` as it does not really belong to the `daemon/config` package. There is still a `pflags` dependency on the `valueSet` field, I added a comment on that to _fix_.

This is the coverage almost _as is_ — just complete the `Validate` test, but the diff is getting pretty big already so I'll do a follow-up for the tests.

```bash
ok      github.com/docker/docker/daemon/config  0.010s  coverage: 60.2% of statements
ok      github.com/docker/docker/daemon/discovery       0.006s  coverage: 31.9% of statements
```

/cc @dnephin @thaJeztah @icecrime @cpuguy83 @crosbymichael @AkihiroSuda @tonistiigi 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

dnephin commented on cmd/dockerd/config_common_unix.go:
```diff
@@ -0,0 +1,29 @@
+
+// installCommonUnixConfigFlags adds command-line options to the top-level flag parser for
+// the current process that are common across Unix platforms.
+func installCommonUnixConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
```
CommonUnix seems like a strange name here. Maybe `installUnixConfigFlags()` ?

dnephin replied:
same goes for the other `CommonUnix` variables

dnephin commented on daemon/config/config.go:
```diff
@@ -20,26 +22,29 @@ import (
 const (
-	defaultShutdownTimeout = 15
+	// DefaultShutdownTimeout is the default shutdown timeout for the daemon
+	DefaultShutdownTimeout = 15
```
This constant appears to be only used in one place. It would be better to keep it unexported and move it to where it is used.

dnephin commented on daemon/config/config_solaris.go:
```diff
@@ -31,17 +31,18 @@ type bridgeConfig struct {
 // InstallFlags adds command-line options to the top-level flag parser for
 // the current process.
-func (config *Config) InstallFlags(flags *pflag.FlagSet) {
+func (conf *Config) InstallFlags(flags *pflag.FlagSet) {
```
Is this still called? It seems like it was removed from the other platforms, so I think this might be broken on solaris.

dnephin reviewed (commented):
Seems like a good incremental step. Still a lot of work to do to make `daemon` a sane package.

vdemeester commented:
@dnephin updated :angel: 

> Still a lot of work to do to make `daemon` a sane package.

You don't say :sweat_smile:

thaJeztah commented:
needs a rebase now @vdemeester :cry:

vdemeester commented:
Hum windows failure is legit.. messed up *build tags* :sweat_smile:

thaJeztah reviewed (approved):
LGTM if green

vdemeester commented:
Rebased 👼

cpuguy83 commented:
Build errors

vdemeester commented:
arf… damn rebase 😂

vdemeester committed:
Extract daemon configuration and discovery to their own package

This also moves some cli specific in `cmd/dockerd` as it does not
really belong to the `daemon/config` package.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

dnephin commented:
needs a rebase

thaJeztah commented:
@vdemeester test is failing on windows;

```
7:56:09 # github.com/docker/docker/cmd/dockerd
17:56:09 cmd\dockerd\config_unix_test.go:22: conf.ShmSize undefined (type *"github.com/docker/docker/daemon/config".Config has no field or method ShmSize)
17:56:09 cmd\dockerd\config_unix_test.go:23: conf.ShmSize undefined (type *"github.com/docker/docker/daemon/config".Config has no field or method ShmSize)
17:56:09 cmd\dockerd\config_unix_test.go:27: conf.ShmSize undefined (type *"github.com/docker/docker/daemon/config".Config has no field or method ShmSize)
17:56:09 cmd\dockerd\config_unix_test.go:28: conf.ShmSize undefined (type *"github.com/docker/docker/daemon/config".Config has no field or method ShmSize)
17:56:09 FAIL	github.com/docker/docker/cmd/dockerd [build failed]
```

vdemeester commented:
/ping @justincormack @runcom @AkihiroSuda 😝

allencloud commented:
need rebase again ^^ @vdemeester

LK4D4 reviewed (approved):
LGTM
